fix(datastore) release startStopSemaphore when start returns, not when API sync completes

richardmcclellan opened the pull request:
- Fixes #1010 ("Unable to acquire orchestrator lock. Transition currently in progress") (also reported in https://github.com/aws-amplify/amplify-android/issues/977#issuecomment-737041548). Previously, if you call two DataStore operations right after each other, the second one would usually fail.  Each operation calls start, which previously required obtaining a semaphore, within a 2 second timeout.  Previously, the semaphore was locked until the API sync completed, but with this PR, the lock is released as soon as the storage observer is initialized (which only takes a few milliseconds).  

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java:
```diff
@@ -122,7 +121,7 @@ public void setup() throws AmplifyException {
     public void itemsPlacedInStorageArePublishedToNetwork() throws AmplifyException {
         // Arrange: orchestrator is running
-        orchestrator.start();
+        orchestrator.start().subscribe();
```
maybe `TestObserver<Void> observer = orchestrator.start().test()`? I think those auto-dispose, which is cool.

jamesonwilliams reviewed (commented):
Regarding thread safety and synchronization of concurrent calls:

If I understand correctly, _all of it_ is now handled by the `Semaphore`? And so we don't need to run the transitions on a `Schedulers.single()`, and we don't need to throw `synchronized` on any method signatures?

richardmcclellan commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -145,169 +143,86 @@ public Orchestrator(
         }
         LOG.debug("Lock acquired.");
-        return true;
+        return completable.doFinally(() -> {
```
I did actually try something like that, and agree the visual symmetry would be nice :). It didn't work though, I was getting timeout's trying to acquire the lock.  I'm pretty sure the lock acquisition has to happen before the subscribing to the `Completable`.

richardmcclellan commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -145,169 +143,86 @@ public Orchestrator(
+        disposables.clear();
+        switch (currentMode.get()) {
             case SYNC_VIA_API:
+            case SYNC_STARTING:
```
I did consider adding the other 3 transitions:
 - `STOPPED` -> `LOCAL_ONLY`
 - `SYNC_VIA_API` -> `LOCAL_ONLY`
 - `LOCAL_ONLY` -> `STOPPED` 

Ultimately I didn't because all of these transitions are synchronized with the semaphore, so there's no benefit of tracking the state.  

I did have to add `SYNC_STARTING` (`LOCAL_ONLY` -> `SYNC_VIA_API`), since this transition is no longer locked with the semaphore, to ensure the API sync wouldn't ever be started multiple times.

richardmcclellan replied:
Looking closer at this, when the state was `SYNC_STARTING` or `SYNC_VIA_API`, I was always doing the same thing, so I don't actually need both.   I've gotten rid of `SYNC_STARTING`, and just updated logic to set the state to SYNC_VIA_API as soon as it starts to sync, rather than at the end of sync.  That way, it will still prevent sync from starting more than once.

richardmcclellan commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -62,7 +60,7 @@
     private final QueryPredicateProvider queryPredicateProvider;
     private final StorageObserver storageObserver;
-    private final Supplier<Mode> targetMode;
+    private final boolean apiSyncEnabled;
```
I was initially a bit confused by `targetMode`.   I first had thought it was the end state for any transition and would only differ from `currentMode` if a transition was in progress, but that is not the case.  `targetMode` never changes.  For example, during a transition from `SYNC_VIA_API` to `LOCAL_ONLY` (due to loss of internet connection), the targetMode is still `SYNC_VIA_API`.  It is more of a _penultimate_ target mode - the desired mode in the absence of errors.  

You do have a good counterpoint though that using an enum instead of a boolean is a little better semantically. 
 `apiSyncEnabled` doesn't seem ambiguous to me, but maybe that's just because I wrote it :).    

Thinking further though - I'm about to add in logic to retry after the orchestrator stops,  However, we only want to retry if it stopped due to an internal error, not if the customer explicitly calls `stop()`.   I'm going to need some state to distinguish between these cases.

Here's what I'm thinking now:
 - Switch back to using the `targetMode` enum, instead the `apiSyncEnabled` boolean I added.
 - When customer calls start(), set targetMode to `LOCAL_ONLY` or `SYNC_VIA_API`.  This would be passed in as a parameter to `Orchestrator.start()`.  
 - `targetMode` will no longer be set on `Orchestrator` initialization.
 - When customer calls stop(), set targetMode to `STOPPED`

Once retry logic is added, the purpose of `targetMode` will be more clear.

richardmcclellan commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -411,7 +344,7 @@ private Completable stopApiSync() {
-     * The mode of operation for the Orchestrator's synchronization logic.
+     * The current state of the Orchestrator.
      */
     public enum Mode {
```
I agree!  I think I'll rename this to `State`, and rename the instances to `currentState` and `targetState`.

richardmcclellan commented on aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java:
```diff
@@ -80,7 +79,7 @@ public void setup() throws AmplifyException {
 
         // SUBSCRIPTIONS_ESTABLISHED indicates that the orchestrator is up and running.
```
The `preventConcurrentStateTransitions` test waits for `orchestratorInitObserver` before verifying that `query` is called once:

```
verify(mockApi, times(1)).query(any(), any(), any());
```

Since we are verifying that the sync query happened, it makes sense to wait for SYNC_QUERIES_READY, rather than SUBSCRIPTIONS_ESTABLISHED.  So, this change is more of a "fix".  My change to not release the semaphore sooner just uncovered the bug.

I will update the comment though!

richardmcclellan commented:
> Regarding thread safety and synchronization of concurrent calls:
>If I understand correctly, all of it is now handled by the Semaphore? And so we don't need to run the transitions on a Schedulers.single(), and we don't need to throw synchronized on any method signatures?

I'm going to take another closer look at this.  _Before_, I think the semaphore handled everything, and the `synchronized` methods were unnecessary.  _Now_, we may actually be able to get rid of the semaphore and just have `synchronized` methods.  I'm going to update the logic for stopping on errors and retrying first though, in my next PR.

richardmcclellan commented:
@jamesonwilliams I think I've addressed all of your comments, so this is ready for review again.  I have another PR coming after this one to fix the "crash when going offline" bug, and also to potentially remove the semaphore completely.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:
```diff
@@ -235,47 +235,38 @@ private Completable initializeStorageAdapter(Context context) {
-                    onComplete::call,
-                    error -> onError.accept(new DataStoreException("Failed to stop DataStore.", error,
-                            "Retry your request."))), onError);
+        waitForInitialization()
```
Should this one happen on `Schedulers.io()` as above?

richardmcclellan replied:
I'm pretty sure it isn't needed since the `waitForInitialization` method switches to `Schedulers.io()`, but I just added it anyway just to be explicit, in case `waitForInitialization` ever changes.

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java:
```diff
@@ -61,6 +62,8 @@
  */
 @RunWith(RobolectricTestRunner.class)
 public final class OrchestratorTest {
+    private static final Logger LOG = Amplify.Logging.forNamespace("amplify:aws-datastore");
```
The existing test tags end with `:test`

jamesonwilliams reviewed (approved):
Nice cleanup!
